### PR TITLE
feat: denommetadata module hook for denom creation and update

### LINKED ISSUE #659: Add `AfterDenomMetadataCreated` hook

### PR DESCRIPTION
**Pull request by srene**
## Description

This PR implements the hooks for the denommetadata module to communicate the denom metadata creation and update events 

----

Closes #659

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow-up issues.

PR review checkboxes:

I have...

- [ ]  Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ]  Targeted PR against the correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to the GitHub issue with discussion and accepted design
- [ ]  Targets only one GitHub issue
- [ ]  Wrote unit and integration tests
- [ ]  Wrote relevant migration scripts if necessary
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)
- [ ]  Updated the scripts for local run, e.g genesis_config_commands.sh if the PR changes parameters
- [ ]  Add an issue in the [e2e-tests repo](https://github.com/dymensionxyz/e2e-tests) if necessary

SDK Checklist
- [ ] Import/Export Genesis
- [ ] Registered Invariants
- [ ] Registered Events
- [ ] Updated openapi.yaml
- [ ] No usage of go `map`
- [ ] No usage of `time.Now()`
- [ ] Used fixed point arithmetic and not float arithmetic
- [ ] Avoid panicking in Begin/End block as much as possible
- [ ] No unexpected math Overflow
- [ ] Used `sendCoin` and not `SendCoins`
- [ ] Out-of-block compute is bounded
- [ ] No serialized ID at the end of store keys
- [ ] UInt to byte conversion should use BigEndian

Full security checklist [here](https://www.faulttolerant.xyz/2024-01-16-cosmos-security-1/)

----;

For Reviewer:

- [ ]  Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  Confirmed all author checklist items have been addressed

---;

After reviewer approval:

- [ ]  In case the PR targets the main branch, PR should not be squash merge in order to keep meaningful git history.
- [ ]  In case the PR targets a release branch, PR must be rebased.

### CODE REVIEW

File: x/denommetadata/proposal_handler.go [hunk 1]
```diff
@@ -21,29 +22,37 @@ func NewDenomMetadataProposalHandler(k types.BankKeeper) govtypes.Handler {
 }
 
 // HandleCreateDenomMetadataProposal is a handler for executing a passed denom metadata creation proposal
-func HandleCreateDenomMetadataProposal(ctx sdk.Context, k types.BankKeeper, p *types.CreateDenomMetadataProposal) error {
+func HandleCreateDenomMetadataProposal(ctx sdk.Context, k *keeper.Keeper, p *types.CreateDenomMetadataProposal) error {
 
 	for _, metadata := range p.TokenMetadata {
-		found := k.HasDenomMetaData(ctx, metadata.Base)
+		found := k.GetBankKeeper().HasDenomMetaData(ctx, metadata.Base)
```
Comment (omritoptix) [inline x/denommetadata/proposal_handler.go:28]: As you already added a keeper I would simply move this logic there vs calling GetBankKeeper()

File: x/denommetadata/types/hooks.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	AfterDenomMetadataUpdate(ctx sdk.Context, metadata banktypes.Metadata) error
+}
+
+var _ DenomMetadataHooks = MultiDenomMetadataHooks{}
+
+// combine multiple DenomMetadata hooks, all hook functions are run in array sequence
+type MultiDenomMetadataHooks []DenomMetadataHooks
+
+// Creates hooks for the DenomMetadata Module.
+func NewMultiRollappHooks(hooks ...DenomMetadataHooks) MultiDenomMetadataHooks {
```
Comment (omritoptix) [inline x/denommetadata/types/hooks.go:20]: Should be called NewMultiMetaDataHooks

File: app/app.go [hunk 3]
```diff
@@ -741,6 +746,9 @@ func New(
 		transferStack.(delayedackmodule.IBCMiddleware),
 	))
 
+	// register the denommetadata hooks
+	app.DenomMetadataKeeper.SetHooks(denommetadatamoduletypes.NewMultiRollappHooks(app.DenomMetadataKeeper.DenomMetadataHooks()))
```
Comment (omritoptix) [inline app/app.go]: this is for other modules who want to register those hooks. so this line is redundant currently as no module want to register hooks (will be added later by @VictorTrustyDev )

Comment (srene) [reply app/app.go]: removed

File: x/denommetadata/types/hooks.go (hunk 2 above)
Comment (srene) [reply x/denommetadata/types/hooks.go]: fixed

File: x/denommetadata/proposal_handler.go (hunk 1 above)
Comment (srene) [reply x/denommetadata/proposal_handler.go]: moved

File: x/denommetadata/keeper/keeper_test.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	ctx := app.GetBaseApp().NewContext(false, tmproto.Header{})
+
+	suite.App = app
+	suite.Ctx = ctx
+}
+
+func (suite *KeeperTestSuite) TestCreateDenom() {
+	keeper := suite.App.DenomMetadataKeeper
+
+	err := keeper.CreateDenomMetadata(suite.Ctx, suite.getTestMetadata())
```
Comment (omritoptix) [inline x/denommetadata/keeper/keeper_test.go:33]: I would add:
1. validation that the denom metadata was actually created in the bank
2. trying to create the same denom and verify I get an error which is already exists

File: x/denommetadata/keeper/keeper_test.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+	suite.Require().NoError(err)
+
+}
+
+func (suite *KeeperTestSuite) TestUpdateDenom() {
+	keeper := suite.App.DenomMetadataKeeper
+
+	err := keeper.UpdateDenomMetadata(suite.Ctx, suite.getTestMetadata())
```
Comment (omritoptix) [inline x/denommetadata/keeper/keeper_test.go:41]: I would add :
1. validation that it was actually updated
2. trying to update a denom that doesn't exist and verify an error

File: x/denommetadata/keeper/keeper_test.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+	suite.Require().Error(err)
+}
+
+func (suite *KeeperTestSuite) getTestMetadata() banktypes.Metadata {
+	return banktypes.Metadata{
+		Name:        "Cosmos Hub Atom",
+		Symbol:      "ATOM",
```
Comment (omritoptix) [inline x/denommetadata/keeper/keeper_test.go:48]: Please change to dym, adym and dymension hub :)

Comment (srene) [reply x/denommetadata/keeper/keeper_test.go]: done

File: x/denommetadata/keeper/keeper_test.go (hunk 4 above)
Comment (srene) [reply x/denommetadata/keeper/keeper_test.go]: done

File: x/denommetadata/keeper/keeper_test.go (hunk 6 above)
Comment (srene) [reply x/denommetadata/keeper/keeper_test.go]: done

File: x/denommetadata/keeper/keeper_test.go (hunk 5 above)
Comment (srene) [reply x/denommetadata/keeper/keeper_test.go]: done

**Review by omritoptix — APPROVED** [review]

**Pull request merged**